pybamm/expression_tree/functions.py

```diff
@@ -3,7 +3,10 @@
 #
 import numbers
 
-import autograd
+try:
+    import autograd
+except ImportError:
+    pass
```

Comment:
It would be better to do something like this (for all such `try-except` blocks -
```py
try:
     import autograd
 except ImportError:
      warnings.warn("Install autograd for ... . See <doc_link> for more details.")
```

Reply:
Another option is to import optional dependencies inside methods only and not as top-level imports in files (there is a mention of this for `matplotlib` in the contributing guide). The guide should be updated too. We can add a new section about how to handle optional dependencies (about both of these things: adding new dependencies, and how to use them and existing ones while adding new code, regardless of whichever way this PR ends up implementing this mechanism).

Reply:
Thanks both, though I was planning to do it this way - https://stackoverflow.com/questions/27361427 
The imports stay top level but with a handling when specific function is called.
The method @Saransh-cpp suggested should give this warning everytime pybamm is imported but I think it would be more ideal to throw a warning when the function with specific optional dependency is called.